### PR TITLE
[executor][consensus] dedup block_tree in Executor

### PR DESCRIPTION
**Pull request by lightmark**
## Motivation

Currently we have 2 block trees in memory, in Consensus and Executor, respectively. They have exactly the same structure (otherwise it would be a bug). But it has the following drawbacks:
1. Code complexity. In fact, one block tree is enough but due to historical reasons, we have two. We have to be careful to maintain the consistency between the two trees.
2. Stateful executor. As part of TCB, executor should be as thin as possible. A heavy executor with a block tree inside definitely complicates a lot of problems.
3. For full nodes, only `state_synchonizer` and `executor` are available. Full nodes will execute and commit transaction linearly so holding a block tree inside `executor` doesn't make sense, which means it should be in Consensus.

This PR dedups the block_tree in Executor and make execution procedure stateless. The only thing stateful is commit procedure which requires the committed blocks and finality proof (ledger_info_with_sigs) matches the current latest committed data.

More cleanups are coming.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

CI

## Related PRs

#939

### CODE REVIEW

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 1]
```diff
@@ -103,10 +101,11 @@ impl<T: Payload> BlockStore<T> {
         max_pruned_blocks_in_mem: usize,
     ) -> BlockTree<T> {
         let (root_block, root_qc, root_li) = (root.0, root.1, root.2);
-
         // root_compute_res will not used anywhere so use default value to simplify code.
         let root_compute_res = StateComputeResult::default();
-        let executed_root_block = ExecutedBlock::new(root_block, root_compute_res);
+        // TODO: check root matches the committed_trees.
+        let root_output = ProcessedVMOutput::new(vec![], state_computer.committed_trees());
```
Comment (zekun000) [inline consensus/src/chained_bft/block_storage/block_store.rs]: TODO: combine root_compute_res with root_output

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 2]
```diff
@@ -118,20 +117,30 @@ impl<T: Payload> BlockStore<T> {
...
+            let parent_trees = tree
+                .get_block(&block.parent_id())
+                .expect("Parent block must exist")
+                .executed_trees()
+                .clone();
+            let (output, state_compute_result) = state_computer
+                .compute(
+                    block.parent_id(),
+                    block.id(),
+                    parent_trees,
```
Comment (zekun000) [inline consensus/src/chained_bft/block_storage/block_store.rs]: we can simplify this to two arguments: parent, current and piggyback the id

File: consensus/src/chained_bft/consensus_types/block.rs [hunk 3]
```diff
@@ -90,17 +92,30 @@ pub struct Block<T> {
...
     /// pending blocks upon restart.
     compute_result: Arc<StateComputeResult>,
 }
 
+impl<T> PartialEq for ExecutedBlock<T>
+where
+    T: Serialize + Default + CanonicalSerialize + PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.block == other.block
```
Comment (zekun000) [inline consensus/src/chained_bft/consensus_types/block.rs]: compare the StateComputeResult too

File: consensus/src/chained_bft/consensus_types/block.rs [hunk 4]
```diff
@@ -440,6 +464,10 @@ where
     pub fn is_nil_block(&self) -> bool {
         self.block().is_nil_block()
     }
+
+    pub fn executed_trees(&self) -> &ExecutedTrees {
```
Comment (zekun000) [inline consensus/src/chained_bft/consensus_types/block.rs]: rename it to sth like `executed_state`

File: execution/executor/src/block_processor.rs [hunk 5]
```diff
@@ -43,23 +40,32 @@ enum Mode {
...
 
-    committed_trees: ExecutedTrees,
+    committed_trees: Arc<Mutex<ExecutedTrees>>,
 
     /// The main block tree data structure that holds all the uncommitted blocks in memory.
-    block_tree: BlockTree<TransactionBlock>,
+    blocks_to_execute: VecDeque<(
+        ExecutableBlock,
+        oneshot::Sender<Result<(ProcessedVMOutput, StateComputeResult)>>,
+    )>,
```
Comment (wqfish) [inline execution/executor/src/block_processor.rs]: comments need update. also i'm not sure why we need to queue these blocks here. probably can get rid of this in another diff.

File: execution/executor/src/block_processor.rs [hunk 6]
```diff
@@ -43,23 +40,32 @@ enum Mode {
...
 
     /// The blocks that are ready to be sent to storage. After pruning `block_tree` we always put
     /// the blocks here before sending them to storage, so in the case when storage is temporarily
     /// unavailable, we will still prune `block_tree` as normal but blocks will stay here for a bit
     /// longer.
-    blocks_to_store: VecDeque<TransactionBlock>,
+    block_batch_to_commit: Option<(CommittableBlockBatch, oneshot::Sender<Result<()>>)>,
```
Comment (wqfish) [inline execution/executor/src/block_processor.rs]: same here. I think we no longer need this thing, if this code becomes pretty much stateless.

File: consensus/src/chained_bft/consensus_types/block.rs (hunk 3 above)
Comment (lightmark) [reply consensus/src/chained_bft/consensus_types/block.rs]: Now I'll stick with @zekun000 's comments. I will followup to see how to fix this.

File: consensus/src/chained_bft/consensus_types/block.rs (hunk 4 above)
Comment (lightmark) [reply consensus/src/chained_bft/consensus_types/block.rs]: Let's do it in another diff.

File: consensus/consensus-types/src/block.rs [hunk 7]
```diff
@@ -95,17 +95,30 @@ pub struct Block<T> {
...
+impl<T> PartialEq for ExecutedBlock<T>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.block == other.block && self.compute_result == other.compute_result
+    }
+}
+
+impl<T> Eq for ExecutedBlock<T> where T: PartialEq {}
```
[the remark this answers is not in the document]
Comment (zamsden) [reply consensus/consensus-types/src/block.rs]: I think you shouldn't implement Eq for a type that only has a PartialEq bound on its parameterized type, so `where T: Eq {}` is better.

**Pull request closed without merging**